back/app/controllers/web_api/v1/projects_controller.rb

```diff
@@ -52,6 +52,70 @@ def index
     )
   end
 
+  # For use with 'Open to participation' homepage widget.
+  # Returns all published or archived projects that are visible to user
+  # and in an active participatory phase (where user can do something).
+  # Ordered by the end date of the current phase, soonest first (nulls last).
+  def index_projects_with_active_participatory_phase
+    # Projects user can see, with active participatory (not information) phase & include the phases.end_at column
+    # We could use the ProjectPermissionSevice step to filter for active phases, but doing it here is more efficient.
+    subquery = policy_scope(Project)
+      .joins('INNER JOIN admin_publications AS admin_publications ON admin_publications.publication_id = projects.id')
+      .where(admin_publications: { publication_status: 'published' })
+      .joins('INNER JOIN phases AS phases ON phases.project_id = projects.id')
+      .where(
+        'phases.start_at <= ? AND (phases.end_at >= ? OR phases.end_at IS NULL) AND phases.participation_method != ?',
+        Time.zone.now.to_fs(:db), Time.zone.now.to_fs(:db), 'information'
+      )
+      .select('projects.*, phases.end_at AS phase_end_at')
+
+    # Perform the SELECT DISTINCT on the outer query
+    @projects = Project
+      .from(subquery, :projects)
+      .distinct
+      .order('phase_end_at ASC NULLS LAST')
+      .preload(phases: { permissions: [:groups] })
+
+    # Projects user can participate in, or where such participation could (probably) be made possible by user
+    # (e.g. user not signed in).
+    # Unfortunately, this breaks the query chain, so we have to start a new one after this.
+    # Also, we will need the action descriptors again later, so we will store them separately.
+
+    # Step 1: Create all action descriptors
```

Comment:
@jamesspeake I'm a bit concerned about this step. I wonder if there's a way to avoid calculating all the action descriptors for all the projects at this point...

For example, say that we're on the first page of pagination, and e.g. the page size is 6. Some projects in our query will have smart groups for certain permissions and some won't. If e.g. the first 6 projects don't have smart groups for any permissions, we can probably assume that we can just return these projects without having to calculate all the action descriptors?

Or if say the first 5 projects have no smart group permissions, but the 6th and 7th do, and the 8th doesn't. In this case we can just calculate the smart group for the first 8 projects.  In that case we will always reach the page size of 6, even if the all projects with smart group permissions ends up excluded.

Not sure if this would work, just thinking out loud here